lib/ranking_function.py

```diff
@@ -0,0 +1,79 @@
+import pandas as pd
+from pyspark.sql.types import Row
+from pyspark.sql import DataFrame
+
+
+def ranking_function(row1: Row, row2: Row) -> int:
+    """Funkcja rankująca zwraca liczbę odpowiadającą podobieństwu dwóch filmów.
+    Punkty są przyznawane jeśli oba filmy posiadają te same wartości
+    w odpowiednich kolumnach.
+
+    :param row1:            pyspark.sql.types.Row
+    :param row2:            pyspark.sql.types.Row
+    """
+    cols = [
+        "czy_dla_doroslych",
+        "rok_wydania_produkcji",
+        "rok_zakonczenia_produkcji",
+        "dlugosc_produkcji_w_min",
+        "liczba_sezonow",
+        "liczba_wszystkich_odcinkow",
+        "rodzaj_produkcji_ohe",
+        "epoka_ohe",
+        "gatunek_vec",
+        "ludzie_filmu",
+    ]
+
+    assert all(x in row1.__fields__ for x in cols)
+    assert all(x in row2.__fields__ for x in cols)
+
+    score = 0
+
+    if row1["czy_dla_doroslych"] == row2["czy_dla_doroslych"]:
+        score += 1
+
+    if abs(row1["rok_wydania_produkcji"] - row2["rok_wydania_produkcji"]) < 10:
+        score += 1
+
+    if abs(row1["rok_zakonczenia_produkcji"] - row2["rok_zakonczenia_produkcji"]) < 10:
+        score += 1
+
+    if abs(row1["dlugosc_produkcji_w_min"] - row2["dlugosc_produkcji_w_min"]) < 15:
+        score += 1
+
+    if row1["liczba_sezonow"] == row2["liczba_sezonow"]:
+        score += 1
+
+    if row1["liczba_wszystkich_odcinkow"] == row2["liczba_wszystkich_odcinkow"]:
+        score += 1
+
+    if row1["rodzaj_produkcji_ohe"] == row2["rodzaj_produkcji_ohe"]:
+        score += 1
+
+    if row1["epoka_ohe"] == row2["epoka_ohe"]:
+        score += 1
+
+    score += row1["gatunek_vec"].dot(row2["gatunek_vec"])
+
+    score += len(set(row1["ludzie_filmu"]) & set(row2["ludzie_filmu"]))
+
+    return score
```

Comment:
Czy udałoby się zrobić z tego funkcję, która nie ma tylu ifów, a tylko jeden for?
```
def ranking_function(row1: Row, row2: Row, cols: Tuple[str, int]) -> int:
      
      for tuple in cols:
            if (abs(row1[tuple[0]] - row2[tuple[0]]) <= tuple[1]:
                  score += 1
```

Zdaję sobie sprawę, że wtedy dodatkowe warunki mogą być znacznie trudniejsze do napisania i wtedy może trzeba by było je odrzucić.

Reply:
Tutaj myślę, że taka opcja nie jest potrzebna. Nasza metryka raczej nie powinna się zmieniać w trakcie obliczeń